lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,625 @@
+  // - Generate a body for the blackbox module with the signal mapping
+
+  // Gather some string attributes in the context to simplify working with the
+  // annotations.
+  Strings strings(&getContext());
+
+  // Gather a list of extmodules that have data or mem tap annotations to be
+  // expanded.
+  SmallVector<AnnotatedExtModule, 4> modules;
+  for (auto &op : *circuitOp.getBody()) {
+    auto extModule = dyn_cast<FExtModuleOp>(&op);
+    if (!extModule)
+      continue;
+
+    // Go through the module ports and collect the annotated ones.
+    AnnotatedExtModule result{extModule, {}, {}};
+    for (unsigned argNum = 0, e = extModule.getNumArguments(); argNum < e;
+         ++argNum) {
+      // Go through all annotations on this port and add the data tap key and
+      // mem tap ones to the list.
+      auto annos = AnnotationSet(
+          extModule.getArgAttrOfType<ArrayAttr>(argNum, strings.fannos));
+      for (auto anno : annos) {
+        if (anno.isClass(strings.memTapClass, strings.deletedKeyClass,
+                         strings.literalKeyClass, strings.referenceKeyClass,
+                         strings.internalKeyClass))
+          result.portAnnos.push_back({argNum, anno});
+      }
+    }
+
+    // If there are data tap annotations on the module, which is likely the
+    // case, create a filtered array of annotations with them removed.
+    if (auto attrs = extModule->getAttrOfType<ArrayAttr>(strings.annos)) {
+      auto isBad = [&](Attribute attr) {
+        if (attr.cast<DictionaryAttr>().getAs<StringAttr>("class") ==
+            strings.dataTapsClass)
+          return true;
+        return false;
+      };
+      bool anyBad = llvm::any_of(attrs, isBad);
+      if (anyBad) {
+        SmallVector<Attribute, 8> filteredAttrs;
+        filteredAttrs.reserve(attrs.size());
+        for (auto attr : attrs) {
+          if (!isBad(attr))
+            filteredAttrs.push_back(attr);
+        }
+        result.filteredModuleAnnos =
+            ArrayAttr::get(&getContext(), filteredAttrs);
+      } else {
+        result.filteredModuleAnnos = attrs;
+      }
+    }
+
+    if (!result.portAnnos.empty())
+      modules.push_back(std::move(result));
+  }
+
+#ifndef NDEBUG
+  for (auto m : modules) {
+    LLVM_DEBUG(llvm::dbgs() << "Extmodule " << m.extModule.getName() << " has "
+                            << m.portAnnos.size() << " port annotations\n");
+  }
+#endif
+
+  // Fast path if there's nothing to do.
+  if (modules.empty()) {
+    markAllAnalysesPreserved();
+    return;
+  }
+
+  // Build an instance graph.
+  InstanceGraph instanceGraph(circuitOp);
+
+  // Gather the annotated ports and operations throughout the design that we are
+  // supposed to tap in one way or another.
+  DenseMap<Attribute, BlockArgument> tappedArgs;
+  DenseMap<Attribute, Operation *> tappedOps;
+  circuitOp.walk([&](Operation *op) {
+    if (auto module = dyn_cast<FModuleOp>(op)) {
```

Comment:
Plz tidy this code up with the new helpers above.

Reply:
Done.